src/data/equiv/derangements/finite.lean

```diff
@@ -0,0 +1,134 @@
+    factorials.
+-/
+
+open equiv fintype perm
+open_locale big_operators
+
+variables {α : Type*} [decidable_eq α] [fintype α]
+
+instance : decidable_pred (@is_derangement α) :=
+begin
+  intro f,
+  apply fintype.decidable_forall_fintype,
+end
+
+instance : fintype (derangements α) :=
+begin
+  have : fintype (perm α) := by apply_instance,
+  dsimp [derangements],
+  exact set_fintype (set_of is_derangement),
+end
+
+/-- The number of derangements on an `n`-element set. -/
+def num_derangements (n : ℕ) : ℕ := card (derangements (fin n))
+
+@[simp] lemma num_derangements_invariant (α : Type*) [fintype α] [decidable_eq α] :
+  card (derangements α) = num_derangements (card α) :=
+begin
+  unfold num_derangements,
+  apply card_eq.mpr,  -- card_eq because we don't need the specific equivalence
+  use derangements_congr (equiv_fin α),
+end
+
+theorem num_derangements_recursive (n : ℕ) :
+  num_derangements (n+2) = (n+1) * num_derangements n + (n+1) * num_derangements (n+1) :=
+begin
+  let X := fin(n+1),
+  -- TODO why can't it infer the `fintype` instances unless i bestow a name on `everything_but`?
+  let everything_but : X → set X := λ a, {a}ᶜ,
+  have card_everything_but : ∀ a : X, card (everything_but a) = n,
+  { intro a,
+    simp only [everything_but, fintype.card_of_finset, set.mem_compl_singleton_iff],
+    rw finset.filter_ne' _ a,
+    rw finset.card_erase_of_mem (finset.mem_univ a),
+    simp },
+  have key := card_congr (@derangements_recursion_equiv X _),
+  rw [num_derangements_invariant, fintype.card_option, fintype.card_fin] at key,
+  simp [card_everything_but, mul_add, key],
+end
+
+lemma num_derangements_0 : num_derangements 0 = 1 := rfl
+
+lemma num_derangements_1 : num_derangements 1 = 0 := rfl
+
+theorem num_derangements_sum (n : ℕ) :
+  (num_derangements n : ℤ) = ∑ k in finset.range (n + 1), (-1 : ℤ)^k * nat.desc_fac k (n - k) :=
+begin
+  refine nat.strong_induction_on n _,
+  clear n, -- to avoid confusion with the n in the hypothesis
+  intros n hyp,
+  -- need to knock out cases n = 0, 1
+  cases n, { rw num_derangements_0, simp },
+  cases n, { rw num_derangements_1, simp [finset.sum_range_succ] },
+  -- now we have n ≥ 2
+  rw num_derangements_recursive,
+  push_cast,
+  -- TODO can these proofs be inferred from some tactic? i tried linarith,
+  -- but for some reason that didn't work
+  rw hyp n (nat.lt_succ_of_le (nat.le_succ _)),
+  rw hyp n.succ (lt_add_one _),
+  -- push all the constants inside the sums, strip off some trailing terms,
+  -- and compare term-wise
+  repeat { rw finset.mul_sum },
+  conv_lhs { congr, skip, rw finset.sum_range_succ },
+  rw [← add_assoc, ← finset.sum_add_distrib],
+  conv_rhs {
+    rw finset.sum_range_succ,
+    rw finset.sum_range_succ,
+    rw add_assoc },
+  -- now both sides are of the form `sum (x : ℕ) in finset.range(n+1) (...) + (...)`, in such
+  -- a way that we can just check that the LHS and RHS match up in each summand
+  congr' 1,
```

Comment:
i feel like this is doable without `conv`? maybe with a `simp only [sum_range_succ]` and then just tidy up a little bit?

Reply:
i tried that once, early on, but i got a lot more extra terms (since we end up with finset.range(n) at the end instead of finset.range(n+1)), and the constant terms got a bit unwieldy. maybe with desc_fac they'll work out though

Reply:
i rearranged things a bit so that the conv blocks are smaller, but because we need to arrange terms in a particular order for `congr'` to work out, i couldn't get rid of them entirely without causing a mess